File: src/main/kotlin/org/kite9/diagram/visualization/orthogonalization/flow/face/Route.java
/**
 * 
 */
package org.kite9.diagram.visualization.orthogonalization.flow.face;

import org.kite9.diagram.common.elements.edge.Edge;
import org.kite9.diagram.visualization.planarization.Face;


/**
 * A route through faces used in the {@link ConstraintGroupGenerator} algorithm
 * 
 * @author robmoffat
 *
 */
class Route {

	Face face;
	public Face getFace() {
		return face;
	}

	int in = -1;
	int out = -1;
	
	public int getIn() {
		return in;
	}
	
	public int getOut() {
		return out;
	}

	public Route getRest() {
		return rest;
	}

	Route rest;
	
	Route(Face f, int in, int out, Route rest) {
		super();
		this.face = f;
		this.rest = rest;
		this.in = in;
		this.out = out;
	}

	public int size() {
		if (rest == null) {
			return 1;
		} else {
			return 1 + rest.size();
		}
	}

	public String toString() {
		if (rest == null) {
			return face.getBoundary(out).toString()+"-"+face.getId()+"-"+face.getBoundary(in).toString();
		} else {
			return face.getBoundary(out).toString()+"-"+face.getId()+"-"+rest.toString();
		}
	}
	
	public boolean containsFace(Face f) {
		return this.face == f ||  ((rest!=null) && (rest.containsFace(f)));
	}
	
	public Edge getInEdge() {
		return face.getBoundary(in);
	}
	
	public Edge getOutEdge() {
		return face.getBoundary(out);
	}
	
}

File: src/main/kotlin/org/kite9/diagram/visualization/orthogonalization/flow/face/Route.kt
/**
 *
 */
package org.kite9.diagram.visualization.orthogonalization.flow.face

import org.kite9.diagram.common.elements.edge.Edge
import org.kite9.diagram.visualization.planarization.Face

/**
 * A route through faces used in the [ConstraintGroupGenerator] algorithm
 *
 * @author robmoffat
 */
data class Route(val face: Face, val _in: Int, val _out: Int, var rest: Route?) {

    fun size(): Int {
        return if (rest == null) {
            1
        } else {
            1 + rest!!.size()
        }
    }

    override fun toString(): String {
        return if (rest == null) {
            face.getBoundary(_out).toString() + "-" + face.getID() + "-" + face.getBoundary(_in).toString()
        } else {
            face.getBoundary(_out).toString() + "-" + face.getID() + "-" + rest.toString()
        }
    }

    fun containsFace(f: Face): Boolean {
        return face === f || rest != null && rest!!.containsFace(f)
    }

    val inEdge: Edge
        get() = face.getBoundary(_in)

    val outEdge: Edge
        get() = face.getBoundary(_out)



}